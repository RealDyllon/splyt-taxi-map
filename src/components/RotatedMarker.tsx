import React, {
  useEffect,
  useRef,
  forwardRef,
  ForwardRefExoticComponent,
  RefAttributes,
} from 'react';
import { Marker, MarkerProps } from 'react-leaflet';
import 'leaflet-rotatedmarker';

interface MyProps extends MarkerProps {
  children: any;
  rotationAngle: number;
  rotationOrigin: any;
}

const RotatedMarker: ForwardRefExoticComponent<
  MyProps & RefAttributes<MarkerProps>
> = forwardRef(
  // eslint-disable-next-line no-shadow
  ({ children, ...props }: MyProps, forwardRef: any) => {
    const markerRef: any = useRef();

    const { rotationAngle, rotationOrigin } = props;
    useEffect(() => {
      const marker: any = markerRef.current;
      if (marker) {
        marker.setRotationAngle(rotationAngle);
        marker.setRotationOrigin(rotationOrigin);
      }
    }, [rotationAngle, rotationOrigin]);

    return (
      <Marker
        ref={(ref) => {
          markerRef.current = ref;
          if (forwardRef) {
            // eslint-disable-next-line no-param-reassign
            forwardRef.current = ref;
          }
        }}
        // eslint-disable-next-line react/jsx-props-no-spreading
        {...props}
      >
        {children}
      </Marker>
    );
  }
);

export default RotatedMarker;
