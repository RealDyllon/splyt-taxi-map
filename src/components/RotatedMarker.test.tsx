import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import { Icon, IconOptions, LatLngTuple } from 'leaflet';

import mapTaxiYellowIcon from '../assets/icons/taxi-yellow-top-down.png';
import driversApiRes from '../data/mock/driversApi';
import RotatedMarker from './RotatedMarker';


const mapTaxiIcon: Icon<IconOptions> = new Icon({
  iconUrl: mapTaxiYellowIcon,
  iconRetinaUrl: mapTaxiYellowIcon,
  popupAnchor: [-0, -0],
  iconSize: [48, 48],
});

const taxiItem = driversApiRes.drivers[0];
const taxiPosition: LatLngTuple = [
  taxiItem.location.latitude,
  taxiItem.location.longitude,
];

const TestComponent = (
  <RotatedMarker
    position={taxiPosition}
    rotationAngle={taxiItem.location.bearing}
    rotationOrigin="center"
    icon={mapTaxiIcon}
    children={()=>{}}
  />
);

it('renders without crashing', () => {
  shallow(TestComponent);
});

it('renders correctly', () => {
  const tree = shallow(TestComponent);
  expect(toJson(tree)).toMatchSnapshot();
});
