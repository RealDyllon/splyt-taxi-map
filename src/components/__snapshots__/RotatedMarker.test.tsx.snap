// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<ForwardRef(ContainerComponent)
  icon={
    NewClass {
      "_initHooksCalled": true,
      "options": Object {
        "iconRetinaUrl": "taxi-yellow-top-down.png",
        "iconSize": Array [
          48,
          48,
        ],
        "iconUrl": "taxi-yellow-top-down.png",
        "popupAnchor": Array [
          -0,
          -0,
        ],
      },
    }
  }
  position={
    Array [
      1.2883802507662878,
      103.8520103955534,
    ]
  }
  rotationAngle={92}
  rotationOrigin="center"
>
  <Component />
</ForwardRef(ContainerComponent)>
`;
